android/src/main/java/com/amplitude/android/Timeline.kt

```diff
@@ -0,0 +1,166 @@
+package com.amplitude.android
+
+import com.amplitude.core.Storage
+import com.amplitude.core.events.BaseEvent
+import com.amplitude.core.platform.Timeline
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.launch
+
+class Timeline : Timeline() {
+    private val eventMessageChannel: Channel<EventQueueMessage> = Channel(Channel.UNLIMITED)
+    var sessionId: Long = -1
+        private set
+    internal var lastEventId: Long = 0
+    var lastEventTime: Long = -1
+
+    internal fun start() {
+        amplitude.amplitudeScope.launch(amplitude.storageIODispatcher) {
+            amplitude.isBuilt.await()
+
+            sessionId = amplitude.storage.read(Storage.Constants.PREVIOUS_SESSION_ID)?.toLong() ?: -1
+            lastEventId = amplitude.storage.read(Storage.Constants.LAST_EVENT_ID)?.toLong() ?: 0
+            lastEventTime = amplitude.storage.read(Storage.Constants.LAST_EVENT_TIME)?.toLong() ?: -1
+
+            for (message in eventMessageChannel) {
+                processEventMessage(message)
+            }
+        }
+    }
+
+    internal fun stop() {
+        this.eventMessageChannel.cancel()
+    }
+
+    override fun process(incomingEvent: BaseEvent) {
+        if (incomingEvent.timestamp == null) {
+            incomingEvent.timestamp = System.currentTimeMillis()
+        }
+
+        eventMessageChannel.trySend(EventQueueMessage(incomingEvent, (amplitude as Amplitude).inForeground))
+    }
+
+    private suspend fun processEventMessage(message: EventQueueMessage) {
+        val event = message.event
+        var sessionEvents: Iterable<BaseEvent>? = null
+        val eventTimestamp = event.timestamp!!
```

Comment:
I believe other event timestamp is bind later in process plugins

Reply:
Not sure what you mean. I believe, timestamps should be assigned as soon as possible to have actual live values Now it is here: https://github.com/amplitude/Amplitude-Kotlin/blob/DXOC-297-DiskReadViolation-in-strict-mode/core/src/main/java/com/amplitude/core/Amplitude.kt#L328.
Other code does fallback assignments like:
```
event.timestamp ?: let {
    event.timestamp = System.currentTimeMillis()
}
```

Reply:
I see, so this is added to cover this. I was worried if this could raiser exceptions.

Reply:
`processEventMessage` is invoked only for items from `eventMessageChannel`
https://github.com/amplitude/Amplitude-Kotlin/blob/DXOC-297-DiskReadViolation-in-strict-mode/android/src/main/java/com/amplitude/android/Timeline.kt#L25
the only way to add items to `eventMessageChannel` is `process` function
https://github.com/amplitude/Amplitude-Kotlin/blob/DXOC-297-DiskReadViolation-in-strict-mode/android/src/main/java/com/amplitude/android/Timeline.kt#L34
here `incomingEvent.timestamp = System.currentTimeMillis()` exists to be ensure to have non-null `timestamp`